pkg/providers/instanceprofile/instanceprofile.go

```diff
@@ -0,0 +1,128 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package instanceprofile
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/aws/aws-sdk-go/service/iam/iamiface"
+	"github.com/mitchellh/hashstructure/v2"
+	"github.com/patrickmn/go-cache"
+	"github.com/samber/lo"
+
+	corev1beta1 "github.com/aws/karpenter-core/pkg/apis/v1beta1"
+	"github.com/aws/karpenter/pkg/apis/settings"
+	"github.com/aws/karpenter/pkg/apis/v1beta1"
+	awserrors "github.com/aws/karpenter/pkg/errors"
+)
+
+type Provider struct {
+	iamapi iamiface.IAMAPI
+	cache  *cache.Cache
+}
+
+func NewProvider(iamapi iamiface.IAMAPI, cache *cache.Cache) *Provider {
+	return &Provider{
+		iamapi: iamapi,
+		cache:  cache,
+	}
+}
+
+func (p *Provider) CreateOrUpdate(ctx context.Context, nodeClass *v1beta1.EC2NodeClass, tags map[string]string) (string, error) {
+	localTags := lo.Assign(tags, map[string]string{v1beta1.LabelNodeClass: nodeClass.Name})
+	profileName := GetProfileName(ctx, nodeClass)
+	delete(localTags, corev1beta1.NodePoolLabelKey)
+
+	// An instance profile exists for this NodeClass and the role name hasn't changed
+	if v, ok := p.cache.Get(fmt.Sprintf("instanceprofile/%s", nodeClass.Name)); ok && v.(string) == nodeClass.Spec.Role {
+		return profileName, nil
+	}
+	// Validate if the instance profile exists and has the correct role assigned to it
+	var instanceProfile *iam.InstanceProfile
+	out, err := p.iamapi.GetInstanceProfileWithContext(ctx, &iam.GetInstanceProfileInput{InstanceProfileName: aws.String(profileName)})
+	if err != nil {
+		if !awserrors.IsNotFound(err) {
+			return "", fmt.Errorf("getting instance profile %q, %w", profileName, err)
+		}
+		o, err := p.iamapi.CreateInstanceProfileWithContext(ctx, &iam.CreateInstanceProfileInput{
+			InstanceProfileName: aws.String(profileName),
+			Tags:                lo.MapToSlice(tags, func(k, v string) *iam.Tag { return &iam.Tag{Key: aws.String(k), Value: aws.String(v)} }),
+		})
+		if err != nil {
+			return "", fmt.Errorf("creating instance profile %q, %w", profileName, err)
+		}
+		instanceProfile = o.InstanceProfile
+	} else {
+		instanceProfile = out.InstanceProfile
+	}
+	if len(instanceProfile.Roles) == 1 {
+		if aws.StringValue(instanceProfile.Roles[0].RoleName) == nodeClass.Spec.Role {
+			return profileName, nil
+		}
+		if _, err = p.iamapi.RemoveRoleFromInstanceProfileWithContext(ctx, &iam.RemoveRoleFromInstanceProfileInput{
+			InstanceProfileName: aws.String(profileName),
+			RoleName:            instanceProfile.Roles[0].RoleName,
+		}); err != nil {
+			return "", fmt.Errorf("removing role %q from instance profile %q, %w", aws.StringValue(instanceProfile.Roles[0].RoleName), profileName, err)
+		}
+	}
+	if _, err := p.iamapi.AddRoleToInstanceProfileWithContext(ctx, &iam.AddRoleToInstanceProfileInput{
```

Comment:
What is the implication if role removal succeeds, but role addition fails? Does this just mean the instance profile has a period of time where it has no role? How does that impact existing instances?